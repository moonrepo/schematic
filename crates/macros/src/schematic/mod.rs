use crate::common::Macro;
use crate::utils::extract_comment;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

pub struct SchematicMacro<'l>(pub Macro<'l>);

impl<'l> ToTokens for SchematicMacro<'l> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let cfg = &self.0;
        let name = cfg.name;
        let schema = cfg.type_of.generate_schema(
            name,
            extract_comment(&cfg.attrs),
            cfg.get_casing_format(),
            cfg.get_tagged_format(),
        );

        tokens.extend(quote! {
            #[automatically_derived]
            impl schematic::Schematic for #name {
                fn generate_schema() -> schematic::SchemaType {
                    use schematic::schema::*;
                    #schema
                }
            }
        });
    }
}
