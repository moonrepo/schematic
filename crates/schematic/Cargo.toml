[package]
name = "schematic"
version = "0.15.1"
edition = "2021"
license = "MIT"
description = "A layered serde configuration and schema library."
homepage = "https://moonrepo.github.io/schematic"
repository = "https://github.com/moonrepo/schematic"
readme = "../../README.md"

[package.metadata.release]
pre-release-replacements = [
	{ file = "../../CHANGELOG.md", search = "Unreleased", replace = "{{version}}" },
]

[package.metadata.docs.rs]
all-features = true

[dependencies]
schematic_macros = { version = "0.15.0", path = "../macros" }
schematic_types = { version = "0.7.0", path = "../types" }
miette = { workspace = true }
thiserror = "1.0.59"
tracing = "0.1.40"

# config
garde = { version = "0.18.0", default-features = false, optional = true, features = [
	"regex",
] }
serde = { workspace = true }
serde_path_to_error = { version = "0.1.16", optional = true }
starbase_styles = { version = "0.4.0", optional = true }

# schema
indexmap = { workspace = true, optional = true, features = ["serde"] }

# json
serde_json = { workspace = true, optional = true, features = [
	"preserve_order",
] }

# json schema
markdown = { version = "1.0.0-alpha.17", optional = true }
schemars = { version = "0.8.17", optional = true, default-features = false }

# toml
toml = { workspace = true, optional = true }

# yaml
serde_yaml = { workspace = true, optional = true }

# url
reqwest = { workspace = true, optional = true, features = ["blocking"] }

[features]
default = ["config", "url"]
config = [
	"dep:garde",
	"dep:serde_path_to_error",
	"dep:starbase_styles",
	"schematic_macros/config",
]
json = ["dep:serde_json"]
schema = ["dep:indexmap", "schematic_macros/schema"]
toml = ["dep:toml"]
url = ["dep:reqwest"]
yaml = ["dep:serde_yaml"]

renderer_json_schema = ["dep:markdown", "dep:schemars", "json", "schema"]
renderer_template = []
renderer_typescript = ["schema"]

type_chrono = ["schematic_types/chrono"]
type_indexmap = ["schematic_types/indexmap"]
type_regex = ["schematic_types/regex"]
type_relative_path = ["schematic_types/relative_path"]
type_rust_decimal = ["schematic_types/rust_decimal"]
type_semver = ["schematic_types/semver"]
type_serde_json = ["schematic_types/serde_json"]
type_serde_toml = ["schematic_types/serde_toml"]
type_serde_yaml = ["schematic_types/serde_yaml"]
type_url = ["schematic_types/url"]

valid_email = ["garde/email"]
valid_url = ["garde/url"]

[dev-dependencies]
schematic = { path = ".", features = [
	"config",
	"json",
	"schema",
	"toml",
	"renderer_json_schema",
	"renderer_template",
	"renderer_typescript",
	"type_chrono",
	"type_indexmap",
	"type_regex",
	"type_relative_path",
	"type_rust_decimal",
	"type_semver",
	"type_serde_json",
	"type_serde_toml",
	"type_serde_yaml",
	"type_url",
	"url",
	"valid_email",
	"valid_url",
	"yaml",
] }
reqwest = { workspace = true, features = [
	"blocking",
	"rustls-tls-native-roots",
] }
serial_test = "3.1.1"
starbase_sandbox = "0.6.0"

# Types
chrono = { workspace = true }
indexmap = { workspace = true }
regex = { workspace = true }
relative-path = { workspace = true, features = ["serde"] }
rust_decimal = { workspace = true }
semver = { workspace = true, features = ["serde"] }
url = { workspace = true, features = ["serde"] }
