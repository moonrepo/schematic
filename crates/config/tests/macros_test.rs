#![allow(dead_code, deprecated)]

use schematic::*;
use serde::{Deserialize, Serialize};
use std::{
    collections::{BTreeMap, HashMap},
    path::PathBuf,
};

fn default_bool<C>(_: &C) -> Option<bool> {
    Some(true)
}

mod private {
    pub fn default_string<C>(_: &C) -> Option<String> {
        Some("bar".into())
    }
}

derive_enum!(
    #[derive(Default, ConfigEnum)]
    pub enum SomeEnum {
        #[default]
        A,
        B,
        C,
    }
);

#[derive(Config)]
pub struct ValueTypes {
    boolean: bool,
    string: String,
    number: usize,
    vector: Vec<String>,
    map: HashMap<String, u64>,
    enums: SomeEnum,
}

#[derive(Config)]
pub struct OptionalValues {
    required: bool,
    optional: Option<String>,
}

#[derive(Config)]
struct DefaultValues {
    #[setting(default = true)]
    boolean: bool,
    #[setting(default = default_bool)]
    boolean_fn: bool,
    // #[setting(default = 'a')]
    // chars: char,
    #[setting(default = "foo")]
    string: String,
    #[setting(default = private::default_string)]
    string_fn: String,
    #[setting(default = "foo.json")]
    file_string: String,
    #[setting(default = "foo with. many values!")]
    long_string: String,
    #[setting(default = "foo/bar")]
    path_string: PathBuf,
    #[setting(default = 123)]
    number: usize,
    #[setting(default = [1, 2, 3, 4])]
    array: [u8; 4],
    #[setting(default = (1, 2, 3, 4))]
    tuple: (u8, u8, u8, u8),
    #[setting(default = vec![1, 2, 3, 4])]
    vector: Vec<usize>,
    // #[setting(default = SomeEnum::default)]
    enums: SomeEnum,
    // Invalid
    // #[setting(default = true, default = default_bool)]
    // invalid: bool,
}

#[derive(Config)]
struct Nested {
    #[setting(nested)]
    one: ValueTypes,
    #[setting(nested)]
    two: Option<ValueTypes>,
    #[setting(nested)]
    list: Vec<ValueTypes>,
    #[setting(nested)]
    map: HashMap<String, ValueTypes>,
    #[setting(nested)]
    map2: Option<BTreeMap<String, ValueTypes>>,
    // Invalid
    // #[setting(nested)]
    // two: bool,
    // #[setting(nested, default = true)]
    // no_defualt: ValueTypes,
}

#[derive(Config)]
#[config(allow_unknown_fields, rename_all = "kebab-case")]
struct Serde {
    #[setting(rename = "renamed")]
    rename: String,
    #[setting(skip)]
    skipped: String,
    #[setting(skip, rename = "renamed")]
    all: bool,
}

#[derive(Config, Serialize)]
#[serde(rename = "SerdeNativeRenamed", rename_all = "snake_case")]
struct SerdeNative {
    #[serde(alias = "test", rename = "renamed")]
    rename: String,
    #[serde(skip)]
    skipped: String,
    #[serde(skip, rename = "renamed")]
    all: bool,
}

fn merge_basic<C>(_: String, _: String, _: &C) -> Result<Option<String>, ConfigError> {
    Ok(None)
}

#[derive(Config)]
struct Merging {
    #[setting(merge = merge_basic)]
    basic: String,
}

#[derive(Config)]
struct ExtendsString {
    #[setting(extend)]
    extends: String,
    // #[setting(extend)]
    // extends2: String,
    // #[setting(extend)]
    // extends_int: usize,
}

#[derive(Config)]
struct ExtendsList {
    #[setting(extend)]
    extends: Vec<String>,
}

#[derive(Config)]
struct ExtendsEnum {
    #[setting(extend)]
    extends: ExtendsFrom,
}

#[derive(Config)]
struct ExtendsOptional {
    #[setting(extend)]
    extends: Option<Vec<String>>,
}

fn vec_from_env(_: String) -> Result<Option<Vec<String>>, ConfigError> {
    Ok(Some(vec![]))
}

#[derive(Config)]
struct EnvVars {
    #[setting(env = "FOO")]
    basic: String,
    #[setting(env = "BAR", parse_env = vec_from_env)]
    advanced: Vec<String>,
    // #[setting(parse_env = vec_from_env)]
    // invalid: Vec<String>,
}

fn validate_test<T, C>(_: &str, _: &T, _: &C) -> Result<(), ValidateError> {
    Ok(())
}

#[derive(Config)]
pub struct NestedValidations {
    #[setting(validate = validate_test)]
    basic: String,
}

#[derive(Config)]
struct Validations {
    #[setting(validate = validate_test)]
    basic: String,
    #[setting(validate = validate_test)]
    optional: Option<String>,
    #[setting(nested)]
    nested: NestedValidations,
    #[setting(nested)]
    nested2: Option<NestedValidations>,
}

/// Container comment.
#[derive(Config)]
#[deprecated(since = "1.2.3", note = "Invalid")]
struct Comments {
    // Normal
    normal: bool,
    /// Docs
    #[deprecated]
    docs: bool,
    /* Inline block */
    inline_block: bool,
    /**
     * Block
     */
    #[deprecated = "Bye"]
    block: bool,
}

#[derive(ConfigEnum, Debug)]
enum BasicEnum {
    Foo,
    /// Comment
    Bar,
    Baz,
}

#[derive(ConfigEnum, Debug, Deserialize, Serialize)]
#[serde(rename = "Test", rename_all = "UPPERCASE")]
enum CustomFormatEnum {
    Foo,
    #[serde(rename = "bAr")]
    Bar,
    #[variant(value = "b-a-z")]
    Baz,
}

#[derive(ConfigEnum, Debug)]
enum OtherEnum {
    Foo,
    #[deprecated]
    Bar,
    Baz,
    #[variant(fallback)]
    Other(String),
}

#[derive(ConfigEnum, Debug, Serialize)]
enum AliasedEnum {
    #[serde(alias = "a")]
    Foo,
    #[serde(alias = "b")]
    Bar,
    #[serde(alias = "c")]
    Baz,
}

#[cfg(feature = "typescript")]
#[test]
fn generates_typescript() {
    use renderers::typescript::*;
    use starbase_sandbox::{assert_snapshot, create_empty_sandbox};

    let sandbox = create_empty_sandbox();
    let file = sandbox.path().join("config.ts");

    let mut generator = schema::SchemaGenerator::default();
    generator.add::<SomeEnum>();
    generator.add::<BasicEnum>();
    generator.add::<CustomFormatEnum>();
    generator.add::<OtherEnum>();
    generator.add::<AliasedEnum>();
    generator.add::<ValueTypes>();
    generator.add::<OptionalValues>();
    generator.add::<DefaultValues>();
    generator.add::<Serde>();
    generator.add::<SerdeNative>();
    generator.add::<Merging>();
    generator.add::<ExtendsString>();
    generator.add::<ExtendsList>();
    generator.add::<ExtendsEnum>();
    generator.add::<ExtendsOptional>();
    generator.add::<EnvVars>();
    generator.add::<NestedValidations>();
    generator.add::<Validations>();
    generator.add::<Comments>();
    // Partials are separate
    generator.add::<PartialDefaultValues>();
    generator.add::<PartialNested>();
    generator.add::<PartialValidations>();
    generator
        .generate(&file, TypeScriptRenderer::default())
        .unwrap();

    assert!(file.exists());
    assert_snapshot!(std::fs::read_to_string(&file).unwrap());

    generator
        .generate(
            &file,
            TypeScriptRenderer::new(TypeScriptOptions {
                const_enum: true,
                enum_format: EnumFormat::Enum,
                object_format: ObjectFormat::Type,
            }),
        )
        .unwrap();

    assert_snapshot!(std::fs::read_to_string(file).unwrap());
}
