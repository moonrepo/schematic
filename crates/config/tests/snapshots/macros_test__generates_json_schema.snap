---
source: crates/config/tests/macros_test.rs
expression: "std::fs::read_to_string(file).unwrap()"
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "PartialValidations",
  "type": "object",
  "properties": {
    "basic": {
      "oneOf": [
        {
          "type": "string"
        },
        {
          "type": "null"
        }
      ]
    },
    "nested": {
      "oneOf": [
        {
          "$ref": "#/definitions/PartialNestedValidations"
        },
        {
          "type": "null"
        }
      ]
    },
    "nested2": {
      "oneOf": [
        {
          "$ref": "#/definitions/PartialNestedValidations"
        },
        {
          "type": "null"
        }
      ]
    },
    "optional": {
      "oneOf": [
        {
          "type": "string"
        },
        {
          "type": "null"
        }
      ]
    }
  },
  "definitions": {
    "AliasedEnum": {
      "type": "string",
      "enum": [
        "foo",
        "bar",
        "baz"
      ]
    },
    "BasicEnum": {
      "type": "string",
      "enum": [
        "foo",
        "bar",
        "baz"
      ]
    },
    "Comments": {
      "title": "Comments",
      "type": "object",
      "required": [
        "block",
        "docs",
        "inlineBlock",
        "normal"
      ],
      "properties": {
        "block": {
          "description": "Block",
          "deprecated": true,
          "type": "boolean"
        },
        "docs": {
          "description": "Docs",
          "deprecated": true,
          "type": "boolean"
        },
        "inlineBlock": {
          "type": "boolean"
        },
        "normal": {
          "type": "boolean"
        }
      }
    },
    "CustomFormatEnum": {
      "type": "string",
      "enum": [
        "FOO",
        "bAr",
        "b-a-z"
      ]
    },
    "DefaultValues": {
      "title": "DefaultValues",
      "type": "object",
      "required": [
        "array",
        "boolean",
        "booleanFn",
        "enums",
        "fileString",
        "longString",
        "number",
        "pathString",
        "string",
        "stringFn",
        "tuple",
        "vector"
      ],
      "properties": {
        "array": {
          "type": "array",
          "items": {
            "type": "number"
          },
          "maxItems": 4,
          "minItems": 4
        },
        "boolean": {
          "type": "boolean"
        },
        "booleanFn": {
          "type": "boolean"
        },
        "enums": {
          "allOf": [
            {
              "$ref": "#/definitions/SomeEnum"
            }
          ]
        },
        "fileString": {
          "type": "string"
        },
        "longString": {
          "type": "string"
        },
        "number": {
          "type": "number"
        },
        "pathString": {
          "type": "string",
          "format": "path"
        },
        "string": {
          "type": "string"
        },
        "stringFn": {
          "type": "string"
        },
        "tuple": {
          "type": "array",
          "items": [
            {
              "type": "number"
            },
            {
              "type": "number"
            },
            {
              "type": "number"
            },
            {
              "type": "number"
            }
          ],
          "maxItems": 4,
          "minItems": 4
        },
        "vector": {
          "type": "array",
          "items": {
            "type": "number"
          }
        }
      }
    },
    "EnvVars": {
      "title": "EnvVars",
      "type": "object",
      "required": [
        "advanced",
        "basic"
      ],
      "properties": {
        "advanced": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "basic": {
          "type": "string"
        }
      }
    },
    "ExtendsEnum": {
      "title": "ExtendsEnum",
      "type": "object",
      "required": [
        "extends"
      ],
      "properties": {
        "extends": {
          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "array",
              "items": {
                "type": "string"
              }
            }
          ]
        }
      }
    },
    "ExtendsList": {
      "title": "ExtendsList",
      "type": "object",
      "required": [
        "extends"
      ],
      "properties": {
        "extends": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    },
    "ExtendsOptional": {
      "title": "ExtendsOptional",
      "type": "object",
      "required": [
        "extends"
      ],
      "properties": {
        "extends": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "type": "string"
              }
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "ExtendsString": {
      "title": "ExtendsString",
      "type": "object",
      "required": [
        "extends"
      ],
      "properties": {
        "extends": {
          "type": "string"
        }
      }
    },
    "Merging": {
      "title": "Merging",
      "type": "object",
      "required": [
        "basic"
      ],
      "properties": {
        "basic": {
          "type": "string"
        }
      }
    },
    "NestedValidations": {
      "title": "NestedValidations",
      "type": "object",
      "required": [
        "basic"
      ],
      "properties": {
        "basic": {
          "type": "string"
        }
      }
    },
    "OptionalValues": {
      "title": "OptionalValues",
      "type": "object",
      "required": [
        "optional",
        "required"
      ],
      "properties": {
        "optional": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "required": {
          "type": "boolean"
        }
      }
    },
    "OtherEnum": {
      "type": "string",
      "enum": [
        "foo",
        "bar",
        "baz"
      ]
    },
    "PartialDefaultValues": {
      "title": "PartialDefaultValues",
      "type": "object",
      "properties": {
        "array": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "type": "number"
              },
              "maxItems": 4,
              "minItems": 4
            },
            {
              "type": "null"
            }
          ]
        },
        "boolean": {
          "oneOf": [
            {
              "type": "boolean"
            },
            {
              "type": "null"
            }
          ]
        },
        "booleanFn": {
          "oneOf": [
            {
              "type": "boolean"
            },
            {
              "type": "null"
            }
          ]
        },
        "enums": {
          "oneOf": [
            {
              "$ref": "#/definitions/SomeEnum"
            },
            {
              "type": "null"
            }
          ]
        },
        "fileString": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "longString": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "number": {
          "oneOf": [
            {
              "type": "number"
            },
            {
              "type": "null"
            }
          ]
        },
        "pathString": {
          "oneOf": [
            {
              "type": "string",
              "format": "path"
            },
            {
              "type": "null"
            }
          ]
        },
        "string": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "stringFn": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "tuple": {
          "oneOf": [
            {
              "type": "array",
              "items": [
                {
                  "type": "number"
                },
                {
                  "type": "number"
                },
                {
                  "type": "number"
                },
                {
                  "type": "number"
                }
              ],
              "maxItems": 4,
              "minItems": 4
            },
            {
              "type": "null"
            }
          ]
        },
        "vector": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "type": "number"
              }
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "PartialNested": {
      "title": "PartialNested",
      "type": "object",
      "properties": {
        "list": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "$ref": "#/definitions/PartialValueTypes"
              }
            },
            {
              "type": "null"
            }
          ]
        },
        "map": {
          "oneOf": [
            {
              "type": "object",
              "additionalProperties": {
                "$ref": "#/definitions/PartialValueTypes"
              },
              "propertyNames": {
                "type": "string"
              }
            },
            {
              "type": "null"
            }
          ]
        },
        "map2": {
          "oneOf": [
            {
              "type": "object",
              "additionalProperties": {
                "$ref": "#/definitions/PartialValueTypes"
              },
              "propertyNames": {
                "type": "string"
              }
            },
            {
              "type": "null"
            }
          ]
        },
        "one": {
          "oneOf": [
            {
              "$ref": "#/definitions/PartialValueTypes"
            },
            {
              "type": "null"
            }
          ]
        },
        "two": {
          "oneOf": [
            {
              "$ref": "#/definitions/PartialValueTypes"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "PartialNestedValidations": {
      "title": "PartialNestedValidations",
      "type": "object",
      "properties": {
        "basic": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "PartialValueTypes": {
      "title": "PartialValueTypes",
      "type": "object",
      "properties": {
        "boolean": {
          "oneOf": [
            {
              "type": "boolean"
            },
            {
              "type": "null"
            }
          ]
        },
        "enums": {
          "oneOf": [
            {
              "$ref": "#/definitions/SomeEnum"
            },
            {
              "type": "null"
            }
          ]
        },
        "map": {
          "oneOf": [
            {
              "type": "object",
              "additionalProperties": {
                "type": "number"
              },
              "propertyNames": {
                "type": "string"
              }
            },
            {
              "type": "null"
            }
          ]
        },
        "number": {
          "oneOf": [
            {
              "type": "number"
            },
            {
              "type": "null"
            }
          ]
        },
        "string": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        },
        "vector": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "type": "string"
              }
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "Serde": {
      "title": "Serde",
      "type": "object",
      "required": [
        "renamed"
      ],
      "properties": {
        "renamed": {
          "type": "string"
        }
      }
    },
    "SerdeNative": {
      "title": "SerdeNative",
      "type": "object",
      "required": [
        "renamed"
      ],
      "properties": {
        "renamed": {
          "type": "string"
        }
      }
    },
    "SomeEnum": {
      "type": "string",
      "enum": [
        "a",
        "b",
        "c"
      ]
    },
    "Validations": {
      "title": "Validations",
      "type": "object",
      "required": [
        "basic",
        "nested",
        "nested2",
        "optional"
      ],
      "properties": {
        "basic": {
          "type": "string"
        },
        "nested": {
          "allOf": [
            {
              "$ref": "#/definitions/NestedValidations"
            }
          ]
        },
        "nested2": {
          "oneOf": [
            {
              "$ref": "#/definitions/NestedValidations"
            },
            {
              "type": "null"
            }
          ]
        },
        "optional": {
          "oneOf": [
            {
              "type": "string"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "ValueTypes": {
      "title": "ValueTypes",
      "type": "object",
      "required": [
        "boolean",
        "enums",
        "map",
        "number",
        "string",
        "vector"
      ],
      "properties": {
        "boolean": {
          "type": "boolean"
        },
        "enums": {
          "allOf": [
            {
              "$ref": "#/definitions/SomeEnum"
            }
          ]
        },
        "map": {
          "type": "object",
          "additionalProperties": {
            "type": "number"
          },
          "propertyNames": {
            "type": "string"
          }
        },
        "number": {
          "type": "number"
        },
        "string": {
          "type": "string"
        },
        "vector": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    }
  }
}

